### PR TITLE
migrate setup.py to pyproject.toml

### PR DESCRIPTION
**Pull request by deronnax**
## Description

Following [PEP 621](https://peps.python.org/pep-0621/), setup.py is soft-deprecated (not truly deprecated but not the way to go anymore) and metadata in pyproject.toml is the new norm. This change should be functionally strictly equivalent.

## Checklist

- [x] I've added this contribution to the `changelog.rst`.
- [x] I've added my name to the `AUTHORS` file (or it's already there).

- [x] I installed pre-commit hooks (`pip install pre-commit && pre-commit install`), and ran `black` on my code.
- [x] Please squash merge this pull request (uncheck if you'd like us to merge as multiple commits)

### CODE REVIEW

**Comment by j-bennet** [comment]
Thanks for the PR @deronnax , this has been overdue for some time. CI is not happy with the change though, do you think you could figure this out, or need help? Our CI workflow is pretty simple, so I'm surprise there's been a hiccup:

https://github.com/dbcli/pgcli/blob/main/.github/workflows/ci.yml

but perhaps we're missing something there.

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,65 @@
...
+    "pgspecial>=2.0.0",
+    "click >= 4.1",
+    "Pygments>=2.0",
+    "prompt_toolkit>=2.0.6,<4.0.0",
+    "psycopg >= 3.0.14; sys_platform != 'win32'",
+    "psycopg-binary >= 3.0.14; sys_platform == 'win32'",
+    "sqlparse >=0.3.0,<0.6",
+    "configobj >= 5.0.6",
+    "cli_helpers[styles] >= 2.2.1",
+    "setproctitle >= 1.1.9",
```
Comment (dbaty) [inline pyproject.toml:36]: This requirement was conditional and there was a helpful comment that explained why. Both (the condition and the comment) should be retained.

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,65 @@
...
+    "Topic :: Software Development",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+]
+urls = { Homepage = "http://pgcli.com" }
+requires-python = ">=3.8"
+dependencies = [
+    "pgspecial>=2.0.0",
+    "click >= 4.1",
+    "Pygments>=2.0",
+    "prompt_toolkit>=2.0.6,<4.0.0",
```
Comment (dbaty) [inline pyproject.toml:30]: There was a comment about the pinning of "prompt_toolkit". If its removal is intentional, I'd rather see it in a distinct commit (which would explain why), that would be clearer.

**Review by dbaty — COMMENTED** [review]
I have left a few comments from a (very) cursory review. I'll make a second pass in the coming days to make sure that the new pyproject-based version is equivalent to what was set in `setup.py`.

File: pyproject.toml (hunk 1 above)
Comment (deronnax) [reply pyproject.toml]: Sorry. Fixed.

File: pyproject.toml (hunk 2 above)
Comment (deronnax) [reply pyproject.toml:32]: Sorry. Fixed as well.

File: pyproject.toml [hunk 3]
```diff
@@ -1,3 +1,71 @@
...
+    "sqlparse >=0.3.0,<0.6",
+    "configobj >= 5.0.6",
+    "cli_helpers[styles] >= 2.2.1",
+    # setproctitle is used to mask the password when running `ps` in command line.
+    # But this is not necessary in Windows since the password is never shown in the
+    # task manager. Also setproctitle is a hard dependency to install in Windows,
+    # so we'll only install it if we're not in Windows.
+    "setproctitle >= 1.1.9; sys_platform != 'win32' and 'CYGWIN' not in sys_platform",
+]
+dynamic = ["version"]
```
[the remark this answers is not in the document]
Comment (deronnax) [reply pyproject.toml:44]: fixed as well.

File: pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,71 @@
...
+
+[project.optional-dependencies]
+keyring = ["keyring >= 12.2.0"]
+sshtunnel = ["sshtunnel >= 0.4.0"]
+
```
[the remark this answers is not in the document]
Comment (deronnax) [reply pyproject.toml:53]: sorry, fixed as well.

**Comment by deronnax** [comment]
All concerns addressed. Sorry for the oversights.

**Comment by deronnax** [comment]
@j-bennet let's merge ?

**Comment by j-bennet** [comment]
@deronnax Yes! thanks for the ping.

**Comment by j-bennet** [comment]
Merged! Well done @deronnax and @dbaty.

**Pull request merged**